pkgs/dart_tooling_mcp_server/lib/src/mixins/analyzer.dart

```diff
@@ -0,0 +1,182 @@
+    });
+    _createAnalysisContext(await listRoots(ListRootsRequest()));
+  }
+
+  /// Creates an analysis context from a list of roots.
+  //
+  // TODO: Better configuration for the DART_SDK location.
+  void _createAnalysisContext(ListRootsResult result) async {
+    final sdkPath = Platform.environment['DART_SDK'];
+    if (sdkPath == null) {
+      throw StateError('DART_SDK environment variable not set');
+    }
+
+    final paths = <String>[];
+    for (var root in result.roots) {
+      var uri = Uri.parse(root.uri);
+      if (uri.scheme != 'file') {
+        throw ArgumentError.value(
+            root.uri, 'uri', 'Only file scheme uris are allowed for roots');
+      }
+      paths.add(p.normalize(uri.path));
+    }
+
+    for (var subscription in _watchSubscriptions) {
+      unawaited(subscription.cancel());
+    }
+    _watchSubscriptions.clear();
+
+    for (var rootPath in paths) {
+      var watcher = DirectoryWatcher(rootPath);
+      _watchSubscriptions.add(watcher.events.listen((event) {
+        final context = _analysisContexts?.contextFor(event.path);
+        if (context == null) return;
+        context.changeFile(p.normalize(event.path));
+      }));
+    }
+
+    _analysisContexts =
+        AnalysisContextCollection(includedPaths: paths, sdkPath: sdkPath);
+  }
+
+  /// Implementation of the [analyzeFilesTool], analyzes the requested files
+  /// under the requested project roots.
+  Future<CallToolResult> _analyzeFiles(CallToolRequest request) async {
+    var contexts = _analysisContexts;
+    if (contexts == null) {
+      return CallToolResult(content: [
+        TextContent(
+            text: 'Analysis not yet ready, please wait a few seconds and try '
+                'again.')
+      ], isError: true);
+    }
+
+    var messages = <TextContent>[];
+    final rootConfigs =
+        (request.arguments!['roots'] as List).cast<Map<String, Object?>>();
+    for (var rootConfig in rootConfigs) {
+      var rootUri = Uri.parse(rootConfig['root'] as String);
+      if (rootUri.scheme != 'file') {
+        return CallToolResult(content: [
+          TextContent(
+              text: 'Only file scheme uris are allowed for roots, but got '
+                  '$rootUri')
+        ], isError: true);
+      }
+      var paths = (rootConfig['paths'] as List?)?.cast<String>();
+      if (paths == null) {
+        return CallToolResult(content: [
+          TextContent(
+              text:
+                  'Missing required argument `paths`, which should be the list '
+                  'of relative paths to analyze.')
+        ], isError: true);
+      }
+
+      var context = contexts.contextFor(p.normalize(rootUri.path));
+      await context.applyPendingFileChanges();
+
+      for (var path in paths) {
+        var normalized =
+            p.normalize(p.isAbsolute(path) ? path : p.join(rootUri.path, path));
+        var errorsResult = await context.currentSession.getErrors(normalized);
```

Comment:
nit: we can consider enabling the prefer finals lint

Reply:
filed https://github.com/dart-lang/ai/issues/46